sync/optimistic.md

```diff
@@ -0,0 +1,290 @@
+	block.slot + SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY <= current_slot
+```
+
+Let only a node which returns `is_optimistic(head) == True` be an *optimistic
+node*. Let only a validator on an optimistic node be an *optimistic validator*.
+
+When this specification only defines behaviour for an optimistic
+node/validator, but *not* for the non-optimistic case, assume default
+behaviours without regard for optimistic sync.
+
+## Mechanisms
+
+## When to optimistically import blocks
+
+A block MUST NOT be optimistically imported, unless either of the following
+conditions are met:
+
+1. The justified checkpoint has execution enabled. I.e.,
+   `is_execution_block(get_block(get_state(head_block).current_justified_checkpoint.root))`
+1. The current slot (as per the system clock) is at least `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` ahead of
+   the slot of the block being imported. I.e., `should_optimistically_import_block(current_slot, block) == True`.
+
+*See [Fork Choice Poisoning](#fork-choice-poisoning) for the motivations behind
+these conditions.*
+
+## How to optimistically import blocks
+
+To optimistically import a block:
+
+- The `execute_payload` function MUST return `True` if the execution
+	engine returns `SYNCING` or `VALID`. An `INVALID` response MUST return `False`.
+- The `validate_merge_block` function MUST NOT raise an assertion if both the
+`pow_block` and `pow_parent` are unknown to the execution engine.
+- The parent of the block MUST NOT have an INVALID execution payload.
+
+In addition to this change to validation, the consensus engine MUST be able to
+ascertain, after import, which blocks returned `SYNCING` and which returned
+`VALID`.
+
+Optimistically imported blocks MUST pass all verifications included in
+`process_block` (withstanding the modifications to `execute_payload`).
+
+A consensus engine MUST be able to retrospectively (i.e., after import) modify
+the status of `SYNCING` blocks to be either `VALID` or `INVALID` based upon responses
+from an execution engine. I.e., perform the following transitions:
+
+- `SYNCING` -> `VALID`
+- `SYNCING` -> `INVALID`
+
+When a block transitions from `SYNCING` -> `VALID`, all *ancestors* of the
+block MUST also transition from `SYNCING` -> `VALID`. Such a block is no longer
+considered "optimistically imported".
+
+When a block transitions from `SYNCING` -> `INVALID`, all *descendants* of the
+block MUST also transition from `SYNCING` -> `INVALID`.
+
+When a block transitions from the `SYNCING` state it is removed from the set of
+`optimistic_roots`.
+
+### Execution Engine Errors
+
+When an execution engine returns an error or fails to respond to a payload
+validity request for some block, a consensus engine:
+
+- MUST NOT optimistically import the block.
+- MUST NOT apply the block to the fork choice store.
+- MAY queue the block for later processing.
+
+### Assumptions about Execution Engine Behaviour
```

Comment:
The other Bellatrix specs don't specify EE assumptions. It makes me wonder if it is better to update the official execution API to reflect this

Reply:
This is a good point, we considered this too. @mkalinin and I spoke about this and ended up deciding to leave this note here and Mikhail would update the EE specs.

Do you have any objections to leaving this here if we also update the EE specs, @terencechain?

Reply:
Here is the respective change to the Engine API https://github.com/ethereum/execution-apis/pull/152

Reply:
> Do you have any objections to leaving this here if we also update the EE specs, @terencechain?

No objection :)